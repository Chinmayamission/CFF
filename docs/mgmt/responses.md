You can view responses by going to the "Response" page from a single form.

![form responses](img/form responses.png)

## Switching views

At the top of the response page is a list of tabs that navigate to available response views. You can click on a tab to show the data represented for that particular view.

## Exporting responsees

You can click the "Download CSV" button to download all responses in the current view as a CSV.

## Response actions

Each row has an "Actions" tab with three buttons: View, Edit, and Inspect.

### Viewing a response

Clicking on "View response" will bring you to a read-only version of the response in the form view. You can switch to editing the response by clicking on the "Edit" button at the top.

![view response](img/view response.png)

### Editing a response

Clicking on "Edit response" allows you to modify the response. Note that if you edit a response and the user has already paid, they will automatically receive a confirmation email.

At the end of the form, a "submit options" section (only available to admins with access to edit responses) lets you select whether you would like to disable this behavior or not -- if you select "no", no confirmation email will be sent. If you select "yes", a confirmation email will be sent only if the user has already paid.

![edit response](img/edit response.png)

### Inspecting a response

Inspecting a response allows you to perform more advanced actions on the response such as sending confirmation emails, entering in payments, or deleting the response.

#### Adding payment history

When inspecting a response, you can add manual payments to the payment history. You can add new payment amounts; this is useful if you would like to use CFF to track manual payments such as cash or check payments.

![response payment](img/response payment.png)

!!!note
    Make sure that each payment you enter has a unique ID.


If you select "Send confirmation email", the user will also have a confirmation email sent to them once the new payment is added. By default, a confirmation email will not be sent once a payment is added.


If you would just like to resend a confirmation email without sending a payment, you can select "Send confirmation email" and then click on the "Send email without adding a payment" button.

!!!warning
    As of now, payments cannot be deleted. If you need a payment to be deleted, consider either adding a payment with a negative amount to cancel out the previous payment, or contact <a href="mailto:webmaster@chinmayamission.com">mailto:webmaster@chinmayamission.com</a> to have the payment removed.

#### Viewing / editing the response value

Click on "Response Value" to view the raw response value. Note that this response value is in a JSON format, and can have nested objects or arrays; this is the raw format of the response as it is actually stored in our system.

You can also directly edit fields by clicking on the green "edit" icon next to each field.

![response value](img/response value.png)

#### Deleting a response

You can delete a response from the "Actions" tab. Note that this action **deletes the response forever**, so be careful if you do decide to do this.

![response delete](img/response delete.png)

#### Response inspector (advanced)

Going to "Inspector (advanced)" lets you inspect the entire structure of the response entry that is stored in our system; this includes not only the raw response value, but information and logs about payments and other metadata. You may not need to use this, but this might be a good place for admins to debug something if there is an issue with the form.

![response inspector](img/response inspector.png)