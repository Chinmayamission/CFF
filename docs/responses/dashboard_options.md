!!! warning
    This feature is still under active development and not complete.

Dashboard options are accessible through the user dashboard view. This is accessible through the `/v2/dashboard/formId` link -- for example, https://forms.beta.chinmayamission.com/v2/dashboard/5c99508834513d000161a237/

## Header

To add a dashboard header, you can insert custom HTML in `dashboardOptions.views.header`. The header will be shown above the tabs.

```
  "dashboardOptions": {
    "header": "<h1>Welcome to the MSC dashboard!</h1><br />"
  }
```

## Adding tabs to the dashboard

You can add a tab to the dashboard by editing the `views` array. Note that each item in `views` must have a unique `id` and a `displayName`.

```
  "dashboardOptions": {
    "views": [
      {
        "id": "basic",
        "displayName": "Edit Basic Info",
        "type": "form",
        "pickFields": [
          "contactName"
        ]
      }
    ]
  }
```

### Template views

The simplest type of view is a template -- this is a HTML template
that displays the response info in the specified template.

For example:
```
{
  "id": "profile",
  "displayName": "View profile",
  "type": "template",
  "template": {
    "html": {
      "Hello, {{value.name.first}} {{value.name.last}}. Here is your info:<br> ....."
    }
  }
}
```

## Form views

The form view shows a subset of a form. You can choose which properties to show using the `pickFields` attribute. Optionally, you can override the uiSchema for those particular fields with the `uiSchema` attribute (the `uiSchema` will merge with the existing uiSchema, but replace the values of overlapping keys).

```json
{
  "id": "subset",
  "displayName": "Edit Basic Info",
  "type": "template",
  "pickFields": [
    "contactName"
  ],
  "uiSchema": {
    "contactName": {
      "classNames": "col-12"
    }
  }
}
```

Right now, we only support top-level fields with `pickFields`.

### Disallow modification of existing items in arrays

You can disallow modification of existing items in an array (but still allow adding additional items) by adding the following `uiSchema` attribute to modify the uiSchema:

```json
{
  "id": "subset",
  "displayName": "Edit Basic Info",
  "type": "template",
  "pickFields": [
    "participants"
  ],
  "uiSchema": {
    "participants": {
      "ui:cff:disableModifExistingItems": true
    }
  }
}
```