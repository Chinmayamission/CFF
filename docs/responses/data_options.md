Data options help manage which columns/views are shown in the response table.

## Changing the columns
If you want to change beyond the default view, add the following to `formOptions`:

```
  "dataOptions": {
    "views": [
      {
        "id": "all",
        "displayName": "All",
        "columns": [
          "ID",
          "DATE_LAST_MODIFIED",
          "DATE_CREATED",
          "COUNTER",
          "email"
        ]
      }
    ]
  }
```

Note that each item in `views` must have an `id` and (optionally) a `displayName`. The `columns` value describes which columns will show up in this view's table view in the "Responses" tab.

### Columns object

To change the title of a column, replace the string in the `columns` array with an object with keys `label` and `value`. For example:

```
  "columns": [
    "ID",
    "DATE_LAST_MODIFIED",
    "DATE_CREATED",
    {"label": "Family ID", "value": "COUNTER"},
    "email"
  ]
```

The columns object also supports a basic level of aggregation. Right now, we support the `calculateLength` parameter, which when set to true, will show the length of the value in the `value` parameter (whether it is a string or an array). For example, to show the number of participants in a response (if participants is an array of objects), use the following header object:

```
{
  "label": "Number of participants",
  "value": "participants",
  "calculateLength": true
}
```

To show the value of a custom expression, set the `queryType` to `expr` and specify the expression value in `queryValue`. For example, this could be useful for showing the amount a user has paid for a particular item by doing a price calculation.

```
{
  "label": "Amount paid for registration",
  "queryType": "expr",
  "queryValue": "10 * (age < 18) + 15 * (age >= 18)"
}
```

To add up matching values in `paymentInfo`. set the `queryType` to `paymentInfoItemPaidSum` and specify the names of payment info items in `queryValue`. For example, this could be useful for showing the amount a user has paid for multiple items (such as item 1 + item 2 + discount).

```
{
  "label": "Amount paid for registration",
  "queryType": "paymentInfoItemPaidSum",
  "queryValue": {
    "names": [
      "Main",
      "Discount"
    ]
  }
}
```

This would match items in `paymentInfo` with `name` equal to "Main" or "Discount" and sum these values. For example, it would be equal to 49.5 for the following value of paymentInfo:

```
[
  {"name": "Main", "amount": 50, "quantity": 1, "total": 50},
  {"name": "Sub", "amount": 10, "quantity": 1, "total": 10},
  {"name": "Discount", "amount": -0.5, "quantity": 1, "total": -0.5}
]
```

Note that if using `paymentInfoItemPaidSum` will also cross-check with amount paid. If a payment is partially paid, then it will reduce the final value accordingly. For example, if the initial sum of paymentInfoItems is equal to 49.5, but if the user has only paid 1/3 of the total amount owed (such as through an installment), the final value will be equal to 49.5 / 3 = 16.5.

If you want to filter only payments in a particular date range, specify both `startDate` and `endDate` in queryValue. Both should be in UTC. This will decrease the amounts shown proportionally as well.

```
{
  "label": "Amount paid for registration",
  "queryType": "paymentInfoItemPaidSum",
  "queryValue": {
    "names": [
      "Main",
      "Discount"
    ],
    "startDate": "2019-01-01T08:00:01.000Z",
    "endDate": "2020-01-01T08:00:00.000Z"
  }
}
```



To run a custom mongodb aggregate query, do the following. The "n" key of the result will end up showing in the column:

```
{
  "label": "Age Group",
  "queryType": "aggregate",
  "queryValue": [
    {
      "$project": {
        "n": {
          "$switch": {
            "branches": [
              {"case": {"$lt": ["$age", 18]}, "then": "Child" },
              {"case": {"$lt": ["$age", 29]}, "then": "CHYK" },
              {"case": {"$lt": ["$age", 41]}, "then": "Setukari" }
            ],
            "default": "Adult"
          }
        }
      }
  ]
}
```

## Adding a summary view
You can add a summary view which runs aggregate queries in MongoDB and then shows the results of those queries in the responses table view. To do so, add an object in dataOptions.views such as the following:
```json
"views": [{
  "id": "summary",
  "displayName": "Summary",
  "type": "stats",
  "stats": [
    {
      "type": "single",
      "title": "Total number of Yajman families",
      "queryType": "aggregate",
      "queryValue": [
        {"$match": {"value.registrationType": "sponsorship"}},
        { "$group": { "_id": null, "n": { "$sum": 1 } } }
      ]
    },
    {
      "type": "single",
      "title": "Total money collected",
      "queryType": "aggregate",
      "queryValue": [{ "$group": { "_id": null, "n": { "$sum": "$amount_paid" } } } ]
    }
  ]
}]
```

### queryType aggregate

When `queryType` is `aggregate`, the queryValue will be calculated by evaluating the first value of `n` in the result set.

### type = "single"

When type is single, this means that a single value will be shown. The "title" attribute will be the title, and the value will be next to it. Make sure that the aggregate result has a key called "n".

### type = "group"

When type is group, this means that a table of values will be shown. The "title" attribute is the title of the table. For example, you can use the following value as an item in `stats`:

```json
{
  "type": "group",
  "title": "Aggregate by city",
  "queryType": "aggregate",
  "queryValue": [
      { "$group": { "_id": "$value.city", "n": { "$sum": 1 } } }
  ]
}
```

## Provide anonymous access

To provide anonymous access to all responses in a form, use the `formOptions.responseListApiKey` parameter. To generate the api key value to set, use this code in Python:

```bash
import uuid
import hashlib
api_key = str(uuid.uuid4())
encoded_api_key = hashlib.sha512(api_key.encode()).hexdigest()
print("api key is ", api_key, "encoded api key is ", encoded_api_key)
```

Then provide the api key to the user using the form, and set `formOptions.responseListApiKey` to the encoded api key.

```bash
{
  "responseListApiKey": "[encoded api key]"
}
```

Then, someone can access the form responses by calling https://drcfbob84gx1k.cloudfront.net/v2/forms/[formId]/responses?apiKey=[apiKey]. Sample JS code (note that a dummy value must be included in the `Authorization` header):

```
fetch("https://drcfbob84gx1k.cloudfront.net/v2/forms/.../responses?dataOptionView=summary&apiKey=...", {headers: {"Authorization": "a"}});
```

### Use api key per dataOptionView

If you want to give anonymous access only to a particular dataOptionView (so that, for example, you only publicly expose certains stats), set the `apiKey` value in the dataOptionView.

```bash
"dataOptions": {
  "views": [
    {
      "id": "testview",
      "name": "test view",
      "apiKey": "[encoded api key]"
    }
  ]
}
```

Then, someone can access the form responses by calling https://drcfbob84gx1k.cloudfront.net/v2/forms/[formId]/responses?dataOptionView=[dataOptionViewId]&apiKey=[apiKey].