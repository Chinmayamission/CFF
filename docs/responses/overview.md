# Response Reporting

Response reporting, display options, and admin aggregation options can all be controlled by the `formOptions.dataOptions` object.

## dataOptions structure

Here is the structure of the `dataOptions` object (all root fields are optional):

```json
{
  "dataOptions": {
    "views": { ... },
    "groups": { ... },
    "export": { ... },
    "search": { ... }
  }
}
```

The **views** field is used to control which tabs show up in the admin responses view.

The **groups** field is used to define groups (such as classes) that individual responses can be assigned to.

The **export** field is used to configure export options to Google Sheets.

The **search** field is used to configure options regarding search endpoint capabilities for responses.