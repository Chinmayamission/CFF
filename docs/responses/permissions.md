# Response permissions

By default, people cannot anonymously view or edit responses by link. If you want to allow people to anonymously view / edit responses with only the link, set the following in formOptions:

```
  "responseCanViewByLink": true,
  "responseCanEditByLink": true
```

## Permissions for login required forms

Note that the above two permissions do not have any effect for forms that require login (forms that have `loginRequired` set to true).
By default, a login required form always prompts users to log in when they go to the form, and it allows users to modify
their existing response by default.

If you want to disable modification of users' existing responses for a login required form, set the following in `formOptions`:

```json
{
  "responseModificationEnabled": false
}
```

If you want to disable submission of users' existing responses altogether for a login required form, set the following in `formOptions`:

```json
{
  "responseSubmissionEnabled": false
}
```