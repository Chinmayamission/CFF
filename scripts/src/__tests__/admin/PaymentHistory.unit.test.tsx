import React from "react";
import { shallow, mount, render } from "enzyme";
import { Provider } from "react-redux";
import PaymentHistory from "../../admin/ResponseTable/ResponseCards/PaymentHistory";
import store from "../../store";
import {
  setResponseDetail,
  onPaymentStatusDetailChange,
  submitNewPayment
} from "../../store/responses/actions";
import { IResponse } from "../../store/responses/types";
import sinon from "sinon";
import MockDate from "mockdate";

describe("PaymentHistory", () => {
  beforeEach(() => {
    // todo: mock date doesn't work
    MockDate.set("7/14/2019");
  });

  afterEach(() => {
    MockDate.reset();
  });

  it("renders table with no payments found", () => {
    let response: IResponse = {
      payment_status_detail: [],
      value: {},
      amount_paid: "0",
      paid: false,
      payment_trail: null,
      paymentInfo: {
        total: 0
      }
    };
    store.dispatch(setResponseDetail(response));
    const wrapper = render(
      <Provider store={store}>
        <PaymentHistory />
      </Provider>
    );
    expect(wrapper.text()).toContain("No rows found");
    // expect(wrapper).toMatchSnapshot();
  });

  it("render several types of payments", () => {
    let response: IResponse = {
      payment_status_detail: [
        {
          amount: "12",
          currency: "USD",
          date: { $date: "date1" },
          id: "id1",
          method: "paypal_ipn"
        },
        {
          amount: "24",
          currency: "USD",
          date: { $date: "date1" },
          id: "id2",
          method: "paypal_ipn"
        }
      ],
      value: {},
      amount_paid: "0",
      paid: false,
      payment_trail: null,
      paymentInfo: {
        total: 0
      }
    };
    store.dispatch(setResponseDetail(response));
    const wrapper = render(
      <Provider store={store}>
        <PaymentHistory />
      </Provider>
    );
    expect(wrapper.text()).toContain("$12.00");
    expect(wrapper.text()).toContain("id1");
    expect(wrapper.text()).toContain("id2");
    expect(wrapper.text()).toContain("$24.00");
    expect(wrapper.text()).not.toContain("No rows found");
    // expect(wrapper).toMatchSnapshot();
  });

  it("Does not let you add a new payment when all fields are not entered", () => {
    let response: IResponse = {
      _id: { $oid: 123 },
      payment_status_detail: [],
      value: {},
      amount_paid: "0",
      paid: false,
      payment_trail: null,
      paymentInfo: {
        total: 0
      }
    };
    store.dispatch(setResponseDetail(response));
    const spy = sinon.spy();
    const wrapper = mount(
      <Provider store={store}>
        <PaymentHistory submitNewPayment={spy} />
      </Provider>
    );
    wrapper.find(".cff-payment-history-btn-add").simulate("click");
    expect(spy.calledOnce).toBe(false);
  });

  it("Lets you add a new payment", () => {
    let response: IResponse = {
      _id: { $oid: 123 },
      payment_status_detail: [],
      value: {},
      amount_paid: "0",
      paid: false,
      payment_trail: null,
      paymentInfo: {
        total: 0
      }
    };
    store.dispatch(setResponseDetail(response));
    store.dispatch(onPaymentStatusDetailChange("amount", "40"));
    store.dispatch(onPaymentStatusDetailChange("method", "unittest_method"));
    store.dispatch(onPaymentStatusDetailChange("id", "4011"));
    const spy = sinon.spy();
    store.dispatch = spy;
    const wrapper = mount(
      <Provider store={store}>
        <PaymentHistory />
      </Provider>
    );
    wrapper.find(".cff-payment-history-btn-add").simulate("click");
    // Todo fix
    // expect(spy.calledOnceWith(submitNewPayment({sendEmail: true}))).toBe(true);
  });
});
