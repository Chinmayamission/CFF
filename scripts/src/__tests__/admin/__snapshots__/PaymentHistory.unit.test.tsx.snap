// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render several types of payments 1`] = `
<div
  class="cff-response-payment-history"
>
  Status: 
  <strong>
    NOT PAID
  </strong>
  <br />
  Amount Owed: 
  <strong>
    $0.00
  </strong>
  <br />
  Amount Paid: 
  <strong>
    $0.00
  </strong>
  <br />
  <br />
  <div
    class="ReactTable"
  >
    <div
      class="rt-table"
      role="grid"
    >
      <div
        class="rt-thead -header"
        style="min-width:400px"
      >
        <div
          class="rt-tr"
          role="row"
        >
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Amount
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Method
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              ID
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Date
            </div>
            <div
              class="rt-resizer"
            />
          </div>
        </div>
      </div>
      <div
        class="rt-tbody"
        style="min-width:400px"
      >
        <div
          class="rt-tr-group"
          role="rowgroup"
        >
          <div
            class="rt-tr -odd"
            role="row"
          >
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              $12.00
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              paypal_ipn
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              id1
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              date1
            </div>
          </div>
        </div>
        <div
          class="rt-tr-group"
          role="rowgroup"
        >
          <div
            class="rt-tr -even"
            role="row"
          >
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              $24.00
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              paypal_ipn
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              id2
            </div>
            <div
              class="rt-td"
              role="gridcell"
              style="flex:100 0 auto;width:100px"
            >
              date1
            </div>
          </div>
        </div>
      </div>
      <div
        class="rt-tfoot"
        style="min-width:400px"
      >
        <div
          class="rt-tr"
          role="row"
        >
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <input
                class="form-control form-control-sm"
                placeholder="Amount"
                value=""
              />
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <select
                class="form-control form-control-sm"
              >
                <option
                  disabled=""
                  selected=""
                  value=""
                >
                  Method
                </option>
                <option
                  value="manual_check"
                >
                  Check
                </option>
                <option
                  value="manual_cash"
                >
                  Cash
                </option>
                <option
                  value="manual_credit_card"
                >
                  Credit Card
                </option>
              </select>
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <input
                class="form-control form-control-sm"
                placeholder="ID"
                value=""
              />
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class=""
                    type="text"
                    value="07/14/2019"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="-loading"
    >
      <div
        class="-loading-inner"
      >
        Loading...
      </div>
    </div>
  </div>
  <div
    class="ccmt-cff-Page-FormPage"
  >
    <form
      autocomplete="false"
      class="rjsf"
    >
      <div
        class="form-group field field-object"
      >
        <div
          class="container-fluid"
        >
          <span />
          <div />
          <div
            class="row"
          >
            <div
              class="form-group field field-boolean"
            >
              <div>
                <label
                  class="control-label"
                />
                <div
                  class="checkbox form-check "
                >
                  <input
                    checked=""
                    class="form-check-input"
                    id="root_sendEmail"
                    style="bottom:0"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="root_sendEmail"
                  >
                    Send confirmation email
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="btn btn-sm btn-primary cff-payment-history-btn-add"
      >
        Add new payment
      </button>
    </form>
  </div>
</div>
`;

exports[`renders table with no payments found 1`] = `
<div
  class="cff-response-payment-history"
>
  Status: 
  <strong>
    NOT PAID
  </strong>
  <br />
  Amount Owed: 
  <strong>
    $0.00
  </strong>
  <br />
  Amount Paid: 
  <strong>
    $0.00
  </strong>
  <br />
  <br />
  <div
    class="ReactTable"
  >
    <div
      class="rt-table"
      role="grid"
    >
      <div
        class="rt-thead -header"
        style="min-width:400px"
      >
        <div
          class="rt-tr"
          role="row"
        >
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Amount
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Method
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              ID
            </div>
            <div
              class="rt-resizer"
            />
          </div>
          <div
            class="rt-th rt-resizable-header -cursor-pointer"
            role="columnheader"
            style="flex:100 0 auto;width:100px"
            tabindex="-1"
          >
            <div
              class="rt-resizable-header-content"
            >
              Date
            </div>
            <div
              class="rt-resizer"
            />
          </div>
        </div>
      </div>
      <div
        class="rt-tbody"
        style="min-width:400px"
      />
      <div
        class="rt-tfoot"
        style="min-width:400px"
      >
        <div
          class="rt-tr"
          role="row"
        >
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <input
                class="form-control form-control-sm"
                placeholder="Amount"
                value=""
              />
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <select
                class="form-control form-control-sm"
              >
                <option
                  disabled=""
                  selected=""
                  value=""
                >
                  Method
                </option>
                <option
                  value="manual_check"
                >
                  Check
                </option>
                <option
                  value="manual_cash"
                >
                  Cash
                </option>
                <option
                  value="manual_credit_card"
                >
                  Credit Card
                </option>
              </select>
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <input
                class="form-control form-control-sm"
                placeholder="ID"
                value=""
              />
            </div>
          </div>
          <div
            class="rt-td"
            role="gridcell"
            style="flex:100 0 auto;width:100px"
          >
            <div>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class=""
                    type="text"
                    value="07/14/2019"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rt-noData"
    >
      No rows found
    </div>
    <div
      class="-loading"
    >
      <div
        class="-loading-inner"
      >
        Loading...
      </div>
    </div>
  </div>
  <div
    class="ccmt-cff-Page-FormPage"
  >
    <form
      autocomplete="false"
      class="rjsf"
    >
      <div
        class="form-group field field-object"
      >
        <div
          class="container-fluid"
        >
          <span />
          <div />
          <div
            class="row"
          >
            <div
              class="form-group field field-boolean"
            >
              <div>
                <label
                  class="control-label"
                />
                <div
                  class="checkbox form-check "
                >
                  <input
                    checked=""
                    class="form-check-input"
                    id="root_sendEmail"
                    style="bottom:0"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="root_sendEmail"
                  >
                    Send confirmation email
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="btn btn-sm btn-primary cff-payment-history-btn-add"
      >
        Add new payment
      </button>
    </form>
  </div>
</div>
`;
